### ISSUE TITLE
Startup performance. 

### ISSUE THREAD

**Issue by Carreau** [ipython/ipython#9908]
I had to use IPython the other days in a bootcamp, on machine that do not have ssd, or 146Gb of Ram, and 36Ghz CPU. IPython took up to **10sec** to start in the command line. I think we should try to make the startup time smaller and see if there are a number of operations that could be delayed to later.

**Comment by gforsyth** [comment]
@Carreau have you played around with @scopatz's amalgamation package yet?   https://github.com/xonsh/amalgamate

https://github.com/xonsh/lazyasd might help also

**Comment by takluyver** [comment]
I've done a bit of work on this years ago - mainly trying to eliminate imports of modules we don't actually need. It's hard to make much difference that way, but we're probably due for another pass. @scopatz's tools look neat, though I'd be a bit worried about the clever things they do with imports breaking other things we haven't thought of.

I wonder how much of that startup time is due to the `pkg_resources` problem where it scans all installed packages on import?

**Comment by gforsyth** [comment]
On our end, is was almost entirely the `pkg_resources` problem

**Comment by scopatz** [comment]
Hi all, my guess is that lazyasd would get IPython most of the way there.  As @takluyver said, I think you probably end up actually using most of your imports in IPython, so really the only thing amalgamate would help with is filesystem thrashing (which can be a big deal on older machines).

Implementing lazily loaded data structures, lazily compiled regexes, etc can have a huge effect on startup time.

> @scopatz's tools look neat, though I'd be a bit worried about the clever things they do with imports breaking other things we haven't thought of.

We have yet to encounter anything that wasn't a couple of line change in xonsh.  There are some annoyances with debugging and amalgamate. We have yet to figure out how to get the traceback to point to the original file.

> I wonder how much of that startup time is due to the pkg_resources problem where it scans all installed packages on import?

> On our end, it was almost entirely the pkg_resources problem

Once the code base is lazy, `pkg_resources` does become the dominating effect.  However, lazyasd has a 'fix' for this by loading it in a background thread. This happens is such a way that if someone actually tries to use it, the calling thread will block until the module is loaded.  See https://github.com/xonsh/lazyasd#background-imports

This is why I think you might be able to mostly get away with lazyasd.  In any event, it is just a single file that you should feel free to drop into IPython if you don't want to have another dependency.  As someone who isn't doing the work, I think it is worth trying out :)

**Comment by willingc** [comment]
@Carreau Do you remember which operating system took 10s at the bootcamp?

**Comment by Carreau** [comment]
Thanks @scopatz and @gforsyth for chiming in. Yes, I've seen both, and want to try this out. 
as amalgamate can also be "opt-in" we might be able to provide that only as a first pass as an option with an env variable.

> Do you remember which operating system took 10s at the bootcamp

The slowest one was an ultramobile windows box laptop, that felt slow even in IE. Though there was also a Old-ish Mac close to 7sec as well.

**Comment by scopatz** [comment]
"opt-in" is a good plan

**Comment by klonuo** [comment]
I just wanted to open another issue about startup timing of new IPython 5. Hope this is the right place.

So I'm on Windows 10, 8GB RAM and i5. The first time I run IPython it can take several seconds, maybe even 10 seconds, like I'm opening Visual Studio.

I tried simply `python -v -m IPython` but could not detect offender - a large list of imports just passes by. So this is only on first time I instantiate IPython, as later it loads quick.

**Comment by takluyver** [comment]
@klonuo in your installed `IPython/__init__.py`, find the line `import sys`, and immediately below, try inserting:

```python
sys.modules['pkg_resources'] = None
```

That should block one particular module we think is slowing it down. Can you see if that makes a noticeable/measurable difference?

**Comment by klonuo** [comment]
Yes, after inserting that line and rebooting (so that I start IPython clean) it starts noticeable quick., w/o need to make any measurements.

**Comment by takluyver** [comment]
OK, so maybe we should look into lazyasd's background module loading.

**Comment by hainm** [comment]
making the startup time much shorter (or using lazy import) will be incredibly helpful. Although It only takes few seconds to go to IPython mode, it's just like 'slow motion' when I just need it for quickly checking function's doc.

**Comment by takluyver** [comment]
I've made PR #10147 with some changes to avoid loading pkg_resources. I didn't use lazyasd for this, because while what it does is really clever, I'm still worried that something will go wrong, and that's the sort of issue that would be horrible to debug.

I made a simpler, more specific change. The downside is that it could be broken if `pygments.plugin` changes, but that doesn't seem a major risk at the present pace of pygments development.

**Comment by o11c** [comment]
Instead of just guessing randomly, I profiled (these specific timings are for 5.1.0, but 6.0.0 is not notably different).

First: real timings, on a hot cache:
```
$ time python3 -m IPython -c 'import sys'
real    0m2.547s
user    0m1.876s
sys     0m0.072s
```
Of this, about half is from `import IPython`. But this comment is about the *other* half.
(as a rule, anything over 0.25 is considered "not for interactive use")

I ran `pprofile` in deterministic mode (*very* slow, so you can't depend on the absolute timings, just the relative ones), opened it in kcachegrind, and found one callstack dominating the post-import time:

```
(uninteresting regex internals snipped since I'm typing this manually)
re.py:22: in compile
pygments/lexer.py:423: in_process_regex
pygments/lexer.py:470: in _process_state
(unknown): <cycle 9>
pygments/lexer.py:514: in process_tokendef
pygments/lexer.py:571: in __call__
prompt_toolkit/layout/lexers.py:174: in __init__
IPython/terminal/ptutils.py:79: in __init__
IPython/terminal/interactiveshell.py:303: in _layout_options
IPython/terminal/interactiveshell.py:209: in init_prompt_toolkit_cli
IPython/terminal/interactiveshell.py:395: in __init__
(and I presume you know how to get there)
```

Or in short: once import is over, it's all `pygments.lexers.Python3Lexer`'s fault. In particular, the grammar is constructed in an absolutely retarded way (speaking as a semi-expert in parsing) - and it certainly doesn't help that python's regexes are perl-based rather than posix-based.

If I had to guess, the biggest win would be dropping all the keyword tokens during regex-time, and doing a dict lookup if it is `name`.

**Comment by takluyver** [comment]
Thanks @o11c for that investigation. Would you have time to have a go at improving the regexes [in pygments](https://bitbucket.org/birkenfeld/pygments-main)? We had some success getting another performance related improvement into pygments.

**Comment by Carreau** [comment]
Would pickling/caching make things faster ? Or maybe we can figure out how to load that lazily onee the user has an already usable interface ?

**Comment by o11c** [comment]
@takluyver I don't have time right now, and it would probably require major internal architectural changes, for which I don't have the experience.

@Carreau pickling wouldn't help, because the pickled form of compiled regexes is just "call `re.compile` with `"abc|def|ghi"` as an argument"

**Comment by RuRo** [comment]
If this ever gets implemented, might I also suggest lazily importing the `pylab` modules? Matplotlib takes around 2 seconds to load on the first import because it's caching various fonts etc. Just lazily loading it or maybe even using `lazyasd.load_module_in_background` should solve this issue.

**Comment by scopatz** [comment]
Yep, those things would help a lot!

**Comment by Asmageddon** [comment]
For me, IPython only loads so slow the first time I run it. I have for now "solved" the issue for me by running `ipython3 -c="exit" ` in my autorun, which makes me think it's probably recompiling all the `.pyc` files.

If this is correct, it could probably be very easily and significantly improved with a solution such as: https://stackoverflow.com/a/2010354

If anyone wants to run some profiling, you can profile import times and visualize them with [tuna](https://pypi.org/project/tuna/):
 `python3 -X importtime $(which ipython3) -c=exit 2>ipython3.import.log;`

**Comment by hainm** [comment]
`ipython3` seems very fast now, took me about 1s. happy to see that improvement.
